### PR TITLE
MetricCallback

### PR DESCRIPTION
**Pull request by TezRomacH**
- Refactor metrics. Now they have to be inherited either from `MetricCallback` or `MultiMetricCallback`
- Remove code duplication
- Apply `yapf` style

### CODE REVIEW

File: utils/metrics.py [hunk 1]
```diff
@@ -19,86 +20,136 @@ def precision(output, target, topk=(1,)):
     return res
 
 
-def apk(actual, predicted, k=10):
+def average_precision(outputs, targets, k=10):
     """
     Computes the average precision at k.
-    This function computes the average prescision at k between two lists of
```
Comment (Scitator) [inline utils/metrics.py:25]: precision?

File: dl/callbacks/metrics.py [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+        self.output_key = output_key
+        self.metric_params = metric_params
+
+    def on_batch_end(self, state):
+        outputs = state.output[self.output_key]
+        targets = state.input[self.input_key]
+
+        metric = self.metric_fn(targets, outputs, **self.metric_params)
+        state.batch_metrics[self.prefix
+                            ] = UtilsFactory.get_val_from_metric(metric)
```
Comment (Scitator) [inline dl/callbacks/metrics.py]: codestyle:
```
state.batch_metrics[self.prefix] = \
    UtilsFactory.get_val_from_metric(metric)
```
?

File: dl/runner.py [hunk 3]
```diff
@@ -9,13 +9,13 @@
 
 from catalyst.utils.factory import UtilsFactory
 from catalyst.utils.misc import merge_dicts
-from catalyst.dl.callbacks import Callback, CheckpointCallback, InferCallback
+from catalyst.dl.callbacks.core import Callback, CheckpointCallback, InferCallback
```
Comment (Scitator) [inline dl/runner.py]: is it possible to import all callbacks just from `catalyst.dl.callbacks`? maybe, add them to __init__?

File: losses/dice.py [hunk 4]
```diff
@@ -1,35 +1,27 @@
-import torch
 import torch.nn as nn
+from catalyst.utils import metrics
```
Comment (Scitator) [inline losses/dice.py]: btw, could we transfer metrics from utils to dl? looks like they finally become DL specific ones
and utils are too complex already

File: losses/focal_loss.py [hunk 5]
```diff
@@ -7,17 +7,18 @@ def __init__(self, gamma=2):
         super().__init__()
         self.gamma = gamma
 
-    def forward(self, input, target):
-        if not (target.size() == input.size()):
+    # TODO refactor
+    def forward(self, outputs, target):
```
Comment (Scitator) [inline losses/focal_loss.py]: targets?

File: utils/metrics.py [hunk 6]
```diff
@@ -19,86 +20,136 @@ def precision(output, target, topk=(1,)):
...
+    Returns
+    -------
+    score : double
+            Dice score
+    """
+    if activation is None or activation == "none":
+        activation_fn = lambda x: x
+    elif activation == "sigmoid":
+        activation_fn = torch.nn.Sigmoid()
+    elif activation == "softmax":
```
Comment (Scitator) [inline utils/metrics.py]: is it better to say softmax2d then? or use `toch.nn.__dict__[name] if name in []`?

**Comment by TezRomacH** [comment]
fixed:
- add `callbacks.core` to `callbacks.__init__` 
- change quotes
- move `metrics` from `catalyst.utils`  to `catalyst.dl`

**Pull request merged**